File: questions/Algorithms/0016. 3Sum Closest/Testing.cc
#include <algorithm>
#include <iostream>
#include <vector>

#include <cc/array.hpp>

using namespace std;

// ------------------------------- solution begin -------------------------------
class Solution {
public:
  int threeSumClosest(vector<int> &nums, int target) {
    sort(nums.begin(), nums.end());
    int ans = nums[0] + nums[1] + nums[nums.size() - 1];
    for (int i = 0; i < nums.size(); i++) {
      int left = i + 1;
      int right = nums.size() - 1;
      while (left < right) {
        int temp = nums[i] + nums[left] + nums[right];
        if (abs(target - ans) > abs(target - temp)) {
          ans = temp;
        }
        if (target - temp > 0) {
          left++;
        } else {
          right--;
        }
      }
    }
    return ans;
  }
};
// ------------------------------- solution end ---------------------------------

int main(int argc, char const *argv[]) {
  vector<int> input1 = {-1, 2, 1, -4};
  int input2 = 1;
  cout << "Input:  " << array_show(input1) << endl;
  Solution solution;
  cout << "Output: " << solution.threeSumClosest(input1, input2) << endl;
  return EXIT_SUCCESS;
}

File: questions/Algorithms/0001. Two Sum/Testing.ts
import array from '../../library/typescript/array';

const twoSum = (nums: number[], target: number): number[] => {
  const map = new Map()

  for (const idx in nums) {
    const diff = target - nums[idx]

    if (map.has(diff)) {
      return [map.get(diff), +idx]
    }

    map.set(nums[idx], +idx)
  }

  return []
}

let input1 = [1, 2, 3];
let input2 = 3;


console.log(`Input:  ${array(input1)}, ${input2}`);
console.log(`Output: ${array(twoSum(input1, input2))}`);
